rust/src/librarydb.rs

```diff
@@ -0,0 +1,504 @@
+    }
+}
+
+impl LibraryDBInner {
+    /// Constructor.
+    fn new() -> Self {
+        Self {
+            arc_libraries: Vec::new(),
+        }
+    }
+
+    /// Opens and adds a library at the end of library database.
+    pub fn add_library(&mut self, data_dir: &Path, library: &Path) -> io::Result<()> {
+        let library = Library::open_library(&data_dir, &library)?;
+        self.arc_libraries.push(Arc::new(RwLock::new(library)));
+        Ok(())
+    }
+
+    /// Opens a library file for reading.
+    /// The file must be dropped before the library database is dropped.
+    pub fn open_file(&self, path: &str) -> io::Result<LibraryFile> {
+        let path = case_insensitive_path(&path);
+        for arc_library in &self.arc_libraries {
+            let library = arc_library.read().unwrap();
+            if let Some(index) = library.find_with_case_insensitive(&path) {
+                return Ok(LibraryFile {
+                    index,
+                    arc_library: arc_library.to_owned(),
+                    position: library.entries[index].data_start,
+                });
+            }
+        }
+        Err(io::ErrorKind::NotFound.into())
+    }
+}
+
+impl Library {
+    /// Opens a library for reading.
+    fn open_library(data_dir: &Path, library: &Path) -> io::Result<Self> {
+        let library_path = find_file(&data_dir, &library)?;
+        let mut library_file = File::open(&library_path)?;
+        let header = SlfHeader::from_input(&mut library_file)?;
+        let base_path = case_insensitive_path(&header.library_path);
+        let mut entries: Vec<_> = header
+            .entries_from_input(&mut library_file)?
+            .iter()
+            .filter_map(|slf_entry| match slf_entry.state {
+                SlfEntryState::Ok => Some(LibraryEntry {
+                    file_path: case_insensitive_path(&slf_entry.file_path),
+                    data_start: u64::from(slf_entry.offset),
+                    data_end: u64::from(slf_entry.offset) + u64::from(slf_entry.length),
+                }),
+                _ => None,
+            }) // LibraryEntry
+            .collect();
+        entries.sort();
+        let mut unique = HashSet::new();
+        for entry in &entries {
+            if !unique.insert(&entry.file_path) {
+                let path = base_path + &entry.file_path;
+                return Err(io::Error::new(
+                    io::ErrorKind::InvalidData,
+                    format!(
+                        "{:?} has multiple entries with path {:?}",
+                        library_path, path
+                    ),
+                ));
+            }
+        }
+        Ok(Library {
+            library_path,
+            library_file,
+            base_path,
+            entries,
+        })
+    }
+
+    /// Finds the file entry that matches the case-insensitive path.
+    /// Returns the entry index or None.
+    fn find_with_case_insensitive(&self, path: &str) -> Option<usize> {
```

Comment:
I would just call this `find_case_insensitive`

Reply:
Renamed it to `find`and clarified what it expects.